File: Lab_8_1.java
import java.util.Arrays;

public class Lab_8_1 {
    public static int coinChange(int[] coins, int amount) {
        // DP массивыг асар том тоогоор инициалжуулна
        int[] dp = new int[amount + 1];
        Arrays.fill(dp, Integer.MAX_VALUE);
        dp[0] = 0;  // Эхний нөхцөл: 0 зоосоор 0 утгыг гаргах боломжтой

        // Бүх зооснуудыг шалгана
        for (int coin : coins) {
            for (int i = coin; i <= amount; i++) {
                // Хэрэв өмнөх зоосоор хийсэн тоо боломжтой бол хамгийн бага утгыг олно
                if (dp[i - coin] != Integer.MAX_VALUE) {
                    dp[i] = Math.min(dp[i], dp[i - coin] + 1);
                }
            }
        }

        // Хэрэв dp[amount] асар том тоотой байвал боломжгүй гэсэн үг, үгүй бол хамгийн бага тоог буцаана
        return dp[amount] == Integer.MAX_VALUE ? -1 : dp[amount];
    }
    public static void main(String[] args) {
        int[] coins = {1, 2, 5};
        int amount = 11;
        System.out.println(coinChange(coins, amount));  // Гаралт: 3
    }
}
// 1. Coin Change Problem (Зоосны солилцооны асуудал):
// (Time Complexity): O(n * amount), энд n нь зоосны тоо.
// (Space Complexity): O(amount), учир нь DP массивыг ашиглаж байгаа.

File: Lab_8_1.py
def coin_change(zoos, hemjee):
    # Тооцоолсон үр дүнг хадгална
    memory = {}

    def dp(uldsen):
        # Үндсэн нөхцөлүүд
        if uldsen == 0:
            return 0  # Хэрэв үлдэгдэл 0 байвал ямар ч мөнгөн дэвсгэрт хэрэггүй
        if uldsen < 0:
            return float('inf')  # Үлдэгдэл сөрөг бол боломжгүй утга

        # Хэрэв өмнө тооцоолсон бол санах ойгоос үр дүнг буцаана
        if uldsen in memory:
            return memory[uldsen]

        # Бүх мөнгөн дэвсгэртийг шалгаж, хамгийн багыг нь сонгоно
        min_zoos = float('inf')
        for coin in zoos:
            ur_dun = dp(uldsen - coin)  # Үлдэгдэлээс дэвсгэртийг хасна
            if ur_dun != float('inf'):
                min_zoos = min(min_zoos, ur_dun + 1)  
                # Хамгийн бага утгыг сонгоно

        # Санах ойд үр дүнг хадгална
        memory[uldsen] = min_zoos
        return min_zoos

    # Гол функцээр дамжуулан нийт үр дүнг тооцоолно
    ur_dun = dp(hemjee)

    # Хэрэв үр дүн хязгааргүй байвал боломжгүй утга гэж үзнэ
    return ur_dun if ur_dun != float('inf') else -1
# Жишээ
zoos = [1, 2, 5]
hemjee = 11
print(coin_change(zoos, hemjee))  # Гаргалт: 3